include/internal/params_template.h

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ * Copyright (c) 2019, Oracle and/or its affiliates.  All rights reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/params.h>
+#include <openssl/ossl_typ.h>
+
+#define OSSL_PARAMTMPL_MAX 10
+
+typedef struct {
+    const char *key;
+    int type;
+    size_t size;
+    size_t alloc_blocks;
+    const BIGNUM *bn;
+    void *string;
+    union {
+        /*
+         * These fields are never directly addressed, but their sizes are
+         * imporant so that all native types can be copied here without overrun.
+         */
+        ossl_intmax_t i;
+        ossl_uintmax_t u;
+        double d;
+    } num;
+} OSSL_PARAMDEF;
+    
+typedef struct {
+    size_t curr;
+    int ispublic;
+    size_t total_blocks;
+    OSSL_PARAMDEF params[OSSL_PARAMTMPL_MAX];
+} OSSL_PARAMTMPL;
```

Comment:
The OSSL_PARAMDEF array has a strong resemblance to a OSSL_PARAM array, and I find it a bit jarring to have a predefined array size like this, so I wonder what you think of this kind of code example:

``` C
    OSSL_PARAM *params;
    OSSL_PARAMDEF tmpl[4];
    size_t tmpl_n = 0;
    OSSL_PARAMDEF_CTX ctx; /* holds calculation data */

    tmpl[tmpl_n++] = ossl_paramtmpl_size_t(&ctx, "size", size);
    tmpl[tmpl_n++] = ossl_paramtmpl_utf8_string(&ctx, "str", s);
    tmpl[tmpl_n++] = ossl_paramtmpl_bn(&ctx, "num", bn);

    params = ossl_paramtmpl_to_param(&ctx, tmpl, tmpl_n, secure_flag);
```

Also, if we don't have the `ctx` at all and calculate the amount of blocks in the `ossl_paramtmpl_to_param` call, it would be enough for those that want to initialize the array as much as possible in compile time and just fill in the values that cannot be set at compile time.

Good for thought.

Reply:
Hmmmmm, it's not as simple anymore but it does remove the arbitrary limit.

Reply:
(side note: that should have been "Food for thought"...)

Reply:
This seems like it is paralleling the OSSL_PARAM array.  Doing what is in #9254 would be be about the same?

Reply:
At the risk of over engineering, a stack of OSSL_PARAMDEFN could be used (with an initial size specified) but that seems like overkill.  This is mean to be simple...

Reply:
> Doing what is in #9254 would be be about the same?

Except that this PR allocates (or uses, if the user provided the block) *one block* for everything.  I like that, all you have to do is free it with `OPENSSL_free`.

Reply:
I think this is a premature optimization. If you're concerned about memory fragmentation, there are doubtless more things in OpenSSL SSL/TLS structures that could be inlined for example.

Reply:
I'm worried about the complexity of cleaning. #9254 means that you can have mixed dynamic and non-dynamic elements in the OSSL_PARAM array, which may turn out to be error prone. This PR is much much cleaner in that regard.

Reply:
I've a plan for freeing memory but this is an all around a cleaner approach.